File: solutions/114_flatten_binary_tree_to_linked_list.cpp
#include "aux.cpp"
/**
Given a binary tree, flatten it to a linked list in-place.
For example, given the following tree:
    1
   / \
  2   5
 / \   \
3   4   6
The flattened tree should look like:
1
 \
  2
   \
    3
     \
      4
       \
        5
         \
          6

 https://leetcode.com/problems/flatten-binary-tree-to-linked-list/description/
 **/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

static const int _ = []() {
  ios::sync_with_stdio(false);
  cin.tie(NULL);
  return 0;
}();
class Solution {
public:
  void flatten(TreeNode *root) {
    if (root != nullptr)
      return;
    flatten(root->left);
    flatten(root->right);
    TreeNode *xright = root->right;
    root->right = root->left;
    root->left = nullptr;

    TreeNode *cur = root;
    while (cur->right)
      cur = cur->right;
    cur->right = xright;
  }
};

int main() {
  Solution s;

  TreeNode *root = new TreeNode(1);
  root->left = new TreeNode(2);
  root->right = new TreeNode(3);
  s.flatten(root);
}

File: _site/solutions/114_flatten_binary_tree_to_linked_list.rb
# Given a binary tree, flatten it to a linked list in-place.
# For example, given the following tree:
#     1
#    / \
#   2   5
#  / \   \
# 3   4   6
# The flattened tree should look like:
# 1
#  \
#   2
#    \
#     3
#      \
#       4
#        \
#         5
#          \
#           6
#
#  https://leetcode.com/problems/flatten-binary-tree-to-linked-list/description/
require './aux.rb'

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Void} Do not return anything, modify root in-place instead.
def flatten2(root)
  return root if root.nil?
  helper = lambda do |r|
    return [nil, nil] if r.nil?
    return [r, r] if r.left.nil? && r.right.nil? # r is never nil
    cleft = helper.call(r.left)
    cright = helper.call(r.right)
    if r.left.nil?
      r.left = nil
      r.right = cright[0]
      return [r, cright.last]
    elsif r.right.nil?
      r.left = nil
      r.right = cleft[0]
      return [r, cleft.last]
    else
      r.left = nil
      cleft[-1].right = cright[0]
      r.right = cleft[0]
      return [r, cright[-1]]
    end
  end
  helper.call(root).first
end

def flatten(root)
  return unless root
  cleft = flatten(root.left)
  cright = flatten(root.right)
  root.left = nil
  root.right = cleft
  cur = root
  cur = cur.right while cur.right
  cur.right = cright
  root
end

arr = [1, 2, 5, 3, 4, nil, 6]
root = construct_tree(arr)
p root
p flatten2(root)
p flatten(root)
